gh-140870: PyREPL auto-complete module attributes in import statements

loic-simon opened the pull request:
Add the ability for `_pyrepl._module_completer.ModuleCompleter` to suggest attributes, in addition to modules.

If the module to import from is not imported yet, ask the user if they want to import it.
This is done by adding the notion of *completion action* to `_pyrepl.completing_reader.CompletingReader`:

  * `get_completions` can return a completion action, in addtion to the suggestions list
  * A completion action is a prompt to show + a callback to call if the user press TAB again
  * The callback can return a message to show instead (here, if import failed)
  * After the callback is fired, `get_completions` is called again, so it can propose new results (here, module attributes)
  * If the user does any other input than TAB when the prompt is shown, the action is discarded

Other changes:
  * I had to tweak a little the code used to display a message below the prompt, to handle the case when it doesn't fit on the console
  *  ~Some unrelated tests to get the coverage rate of `_module_completer` to 100% (in a separate commit, I can remove it if needed)~ -> https://github.com/python/cpython/pull/143244

Interactive WebAssembly demo (up to date): https://pyrepl-attributes-import-completion.pages.dev

Possible improvement: make modules printing stuff when imported not mess with the prompt ; see [this Discourse message](https://discuss.python.org/t/looking-for-feedback-on-adding-import-autocomplete-to-pyrepl/82281/52)

<!-- gh-issue-number: gh-140870 -->
* Issue: gh-140870
<!-- /gh-issue-number -->

loic-simon committed:
gh-140870 PyREPL: add import completion for attributes

loic-simon committed:
Improve behavior when both submodules + prompt

loic-simon committed:
Auto-import stdlib modules

loic-simon committed:
Blacklist some stdlib modules from auto-import

loic-simon committed:
Fix tests post rebase

loic-simon committed:
Improve & test auto-import blacklist

loic-simon committed:
Add _ios_support to auto import blacklist

loic-simon committed:
Fix tests requiring tkinter

StanFromIreland commented on Lib/_pyrepl/_module_completer.py:
```diff
@@ -27,6 +30,17 @@
     "xml.parsers.expat": ["errors", "model"],
 }
 
+AUTO_IMPORT_BLACKLIST = {
```
Can we please rename this? See the [devguide](https://devguide.python.org/documentation/style-guide/#charged-terminology-to-avoid) for suggestions.

loic-simon replied:
Done, I wasn't aware of this, thanks!

loic-simon committed:
Avoid charged terminology

loic-simon committed:
Use always available modules in tests + invalidate caches

loic-simon committed:
Don't test with html module either

loic-simon committed:
Fix refleaks in test_pyrepl.py

pablogsal committed:
Fix pyrepl import completion edge cases

Tighten the pyrepl completer around a few import-completion edge cases.
Relative imports now share one resolver, find_spec failures no longer
bubble out, failed-import tracking resets when sys.path changes, and
double-tab import actions are cleared once consumed.

The patch also keeps empty message lines intact in reader output and
updates the targeted tests and comments to match the intended behavior.

pablogsal reviewed (approved):
LGTM

loic-simon committed:
Merge branch 'main' into pyrepl-import-completion-for-attributes
